observatory-dags/observatory/dags/telescopes/crossref_events.py

```diff
@@ -0,0 +1,388 @@
+            batch_start_date = dt.strftime("%Y-%m-%d")
+            # if final batch or end date is reached before end of period, use end date of release instead
+            if i == (max_processes - 1) or dt == end_date:
+                batch_end_date = end_date.strftime("%Y-%m-%d")
+                batches.append((batch_start_date, batch_end_date))
+                break
+            # use end date of period
+            else:
+                # end date is included, so subtract 1 day from batch_no_days
+                batch_end_date = (dt + timedelta(days=batch_no_days - 1)).strftime("%Y-%m-%d")
+                batches.append((batch_start_date, batch_end_date))
+
+        return batches
+
+    @property
+    def urls(self):
+        urls = []
+        for batch in self.batch_dates:
+            start_date = batch[0]
+            end_date = batch[1]
+
+            self.events_url = f'https://api.eventdata.crossref.org/v1/events?mailto={CrossrefEventsTelescope.MAILTO}' \
+                              f'&from-collected-date={start_date}&until-collected-date={end_date}&rows=1000'
+            self.edited_url = f'https://api.eventdata.crossref.org/v1/events/edited?' \
+                              f'mailto={CrossrefEventsTelescope.MAILTO}&from-updated-date={start_date}' \
+                              f'&until-updated-date={end_date}&rows=1000'
+            self.deleted_url = f'https://api.eventdata.crossref.org/v1/events/deleted?' \
+                               f'mailto={CrossrefEventsTelescope.MAILTO}&from-updated-date={start_date}' \
+                               f'&until-updated-date={end_date}&rows=1000'
+
+            event_type_urls = [self.events_url]
+            if not self.first_release:
+                event_type_urls.append(self.edited_url)
+                event_type_urls.append(self.deleted_url)
+            urls.append(event_type_urls)
+        return urls
+
+    def batch_path(self, url, cursor: bool = False) -> str:
+        """ Gets the appropriate file path for a single batch, either for an events or cursor file.
+
+        :param url: The url used for a specific batch
+        :param cursor: Whether this is a cursor file or file with actual events
+        :return: Path to the events or cursor file
+        """
+        event_type = url.split('?mailto')[0].split('/')[-1]
+        if event_type == 'events':
+            batch_start = url.split('from-collected-date=')[1].split('&')[0]
+            batch_end = url.split('until-collected-date=')[1].split('&')[0]
+        else:
+            batch_start = url.split('from-updated-date=')[1].split('&')[0]
+            batch_end = url.split('until-updated-date=')[1].split('&')[0]
+
+        if cursor:
+            return os.path.join(self.download_folder, f'{event_type}_{batch_start}_{batch_end}_cursor.txt')
+        else:
+            return os.path.join(self.download_folder, f'{event_type}_{batch_start}_{batch_end}')
```

Comment:
Perhaps this should have a file type of `.json`?

Reply:
Mmm, yes it did not mach up with the file type. I decided to keep it a .txt, but instead change the way the cursor was written to the file, so it is actually a text file. There was no need to use json.